abad.go

```diff
@@ -97,4 +121,103 @@ func (a *Abad) evalUnaryExpr(expr *ast.UnaryExpr) (Obj, error) {
 	}
 
 	return num, nil
+}
+
+func (a *Abad) evalExpr(n ast.Node) (types.Value, error) {
+	if !ast.IsExpr(n) {
+		panic("internal error: not an expression")
+	}
+
+	switch n.Type() {
+	case ast.NodeNumber:
+		val := n.(ast.Number)
+		return types.Number(val.Value()), nil
+	case ast.NodeIdent:
+		val := n.(ast.Ident)
+		return a.evalIdentExpr(val)
+	case ast.NodeMemberExpr:
+		val := n.(*ast.MemberExpr)
+		return a.evalMemberExpr(val)
+	case ast.NodeCallExpr:
+		val := n.(*ast.CallExpr)
+		return a.evalCallExpr(val)
+	case ast.NodeUnaryExpr:
+		expr := n.(*ast.UnaryExpr)
+		return a.evalUnaryExpr(expr)
+	}
+
+	panic("unreachable")
+	return nil, nil
+}
+
+func (a *Abad) evalIdentExpr(ident ast.Ident) (types.Value, error) {
+	val, err := a.global.Get(utf16.Str(ident))
+	if err != nil {
+		return nil, err
+	}
+
+	if types.StrictEqual(val, types.Undefined) {
+		return nil, fmt.Errorf("%s is not defined",
+			ident.String())
+	}
+
+	return val, nil
+}
+
+func (a *Abad) evalMemberExpr(member *ast.MemberExpr) (types.Value, error) {
+	objval, err := a.evalExpr(member.Object)
+	if err != nil {
+		return nil, err
+	}
+
+	if objval.Kind() != types.KindObject {
+		panic("wrapping primitive values not implemented yet")
+	}
+
+	obj, err := objval.ToObject()
+	if err != nil {
+		return nil, err
+	}
+
+	return obj.Get(utf16.Str(member.Property))
+}
+
+func (a *Abad) evalCallExpr(call *ast.CallExpr) (types.Value, error) {
+	// TODO(i4k): safe to assume the AST is ok?
+	objval, err := a.evalExpr(call.Callee)
+	if err != nil {
+		return nil, err
+	}
+
+	obj, err := objval.ToObject() // wraps primitives (if needed)
+	if err != nil {
+		return nil, err
+	}
+
+	fun, ok := obj.(types.Function)
```

Comment:
Hmm doesn't JS have a concept of "callable" ? Like even if something is not a function is can be "called" like x() ? (there is in Python, not sure of JS).

Reply:
yes, it has. Functions in JS are just 'internal objects' with a Call method (and usually with a Code property when user function). But it's an internal convention (not exposed to JS code). As we have a type specific to Function in abad, do you think makes sense to abstract into callable ? I mean, use type-switch in the callable interface?

Reply:
What I meant is something else that is not a function be called. Like in Python you can instantiate an object and call it, like:

```python
a = SomeObject()
a()
```

Does JS do something like this ? If not I think the interface is not necessary.

Reply:
I'm not sure, but I think this is not possible. If something is callable I think it should be a Function.